newsrevenuehub/rev-engine
# DEV-3335: Add upgrade prompt to donations page

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

n/a

#### What's this PR do?

- Adds an upgrade prompt to the Contributions page.
- Refactors tracking of when users closed an upgrade prompt into a reusable hook.
- Adds a `<Link>` component which is verbatim from https://github.com/newsrevenuehub/rev-engine/pull/1040
- Renames `<CoreUpgradePrompt>` from https://github.com/newsrevenuehub/rev-engine/pull/1047 to `<SidebarCoreUpgradePrompt>` to avoid naming confusion with this other upgrade prompt.

#### Why are we doing this? How does it help us?

Gets us closer to implementing Core upgrades.

#### How should this be manually tested? Please include detailed step-by-step instructions.

To test the basic workflow:

1. Create a new account, but do not connect Stripe.
2. Confirm that the prompt doesn't appear when you go to the Contributions section. Log out of your session and back in to be sure.
3. Connect your account to Stripe.
4. Confirm that now the prompt appears when you go to the Contributions section.
5. Confirm that clicking on the Learn More button takes you to the pricing page in a new tab, and that nothing happens in the app itself.
6. Log out of your session and back in and confirm the prompt is still there.
7. Close the prompt.
8. Navigate around the app and confirm the prompt doesn't show up again.
9. Hit reload in your browser and confirm the prompt doesn't show up.
10. Log out of your session and log back in.
11. Confirm that the prompt is back.

To test other types of users not seeing it:

1. Log in as a Hub admin. Confirm the prompt doesn't show up.
2. Log in as a user who is on the Plus plan. Confirm the prompt doesn't show up.
3. If possible, log in as a user on the Core plan. Confirm the prompt doesn't show up.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

We may want to document the prompt.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

I had to make a new hook for storing state in session storage. There are several existing ones in NPM that I documented in the file, but none say they are compatible with the version of React we're using. They probably would work anyway, but this requires changing npm install commands to override the peer dependency specified.

I positioned the upgrade prompt using absolute positioning instead of trying to use the `<Hero>` component, which hard-codes controls in the upper-right corner. We will probably want to revisit the component when we have more flexible needs there.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-3335

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.
